### DISCUSSION TITLE
How to onDelete: Cascade on mandatory field

### DISCUSSION THREAD

**Discussion by andrecasal** [prisma/prisma#18007] [Q&A]
I have this `User` model with a 1-to-1 relation to `Profile`:
```prisma
model User {
   id           String          @id @default(uuid())
   createdAt    DateTime        @default(now())
   updatedAt    DateTime        @updatedAt
   email        String          @unique
   passwordHash String
   profile      Profile         @relation(fields: [profileId], references: [id])
   profileId    String          @unique
}

model Profile {
   id          String  @id @default(uuid())
   firstname   String?
   lastname    String?
   user        User?
}
```
The `@relation` is on the `User`model so that we can't make the User creatable without also creating a Profile.

But how do I make the user deletes cascade to profiles?

**Comment by ludralph** [comment]
Hi @andrecasal 👋 

To onDelete cascade, you can add the onDelete argument to the `@relation` directive in the User model. 

```js
model User {
   id           String          @id @default(uuid())
   createdAt    DateTime        @default(now())
   updatedAt    DateTime        @updatedAt
   email        String          @unique
   passwordHash String
   profile      Profile         @relation(fields: [profileId], references: [id], onDelete: Cascade) // notice the onDelete Cascade
   profileId    String          @unique
}

model Profile {
   id          String  @id @default(uuid())
   firstname   String?
   lastname    String?
   user        User?
}
```
This implies that if the `Profile` is deleted, then the `User` associated with that `Profile` will also be deleted.
For example, 
The code below creates a User and a Profile for that User
```js
const res = await prisma.user.create({
    data: {
      email: 'user@example.com',
      passwordHash: 'hashed_password',
      profile: {
        create: {
          firstname: 'John',
          lastname: 'Doe'
        }
      }
    }
  });
```
The output is 

```js
{
  id: '76d8d919-5819-42bf-ae0a-1b2b35082a50',
  createdAt: 2023-02-20T18:01:10.107Z,
  updatedAt: 2023-02-20T18:01:10.107Z,
  email: 'user@example.com',
  passwordHash: 'hashed_password',
  profileId: 'cedf35c5-e817-4bcb-a0c5-e2e973c8d236'
}
```
Now let's get all Users with Profile
```js
  const res = await prisma.user.findMany({
    include:{
      profile: true
    }
   })
````
The output is 
```js
[
  {
    id: '76d8d919-5819-42bf-ae0a-1b2b35082a50',
    createdAt: 2023-02-20T18:01:10.107Z,
    updatedAt: 2023-02-20T18:01:10.107Z,
    email: 'user@example.com',
    passwordHash: 'hashed_password',
    profileId: 'cedf35c5-e817-4bcb-a0c5-e2e973c8d236',
    profile: {
      id: 'cedf35c5-e817-4bcb-a0c5-e2e973c8d236',
      firstname: 'John',
      lastname: 'Doe'
    }
  }
]
```
Let's now delete the Profile
```js

   const res = await prisma.profile.delete({
    where:{
      id: 'cedf35c5-e817-4bcb-a0c5-e2e973c8d236'
    }
   })
```
We can now confirm that both the Profile and the User have been deleted because we get an [ ] when we run
```js
await prisma.user.findMany()
```
and 
```js
await prisma.profile.findMany()
````

**Reply by andrecasal** [reply]
Hey @ludralph 👋

Thank you for your reply! What I was looking for was to cascade the deletes from user to profile instead. So if I delete a User, the associated profile would also be deleted 😊

**Reply by ludralph** [reply]
In that case, you would need to invert the relationship.

```js
model User {
   id           String          @id @default(uuid())
   createdAt    DateTime        @default(now())
   updatedAt    DateTime        @updatedAt
   email        String          @unique
   passwordHash String
   profile      Profile?
}

model Profile {
   id          String  @id @default(uuid())
   firstname   String?
   lastname    String?
   user        User  @relation(fields: [userId], references: [id], onDelete: Cascade)
   userId String @unique
}
```

**Reply by andrecasal** [reply]
Making sure a Profile has to be created with a User is a requirement. Wouldn't this allow for a User to be created without one, because there's no profileId field in the User table?

**Reply by andrecasal** [reply]
This is the current scenario 🤔:
- The creation of a `User` should always involve the creation of a corresponding `Profile`, for which a `profileId` and an `@relation` on the `User` are necessary.
- When a `User` is deleted, its related `Profile` should be automatically deleted as well.
- However, it is not possible to establish @relations on both models using Prisma.

I'm guessing that to `onDelete: Cascade` a `User` deletion to its `Profile`, Prisma would need a `userId` on the `Profile`, which would be impossible because one of the models needs to be created first.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
